Replace priority scheme negotiation (#112)
kazuho/draft-kazuho-httpbis-priority: draft-kazuho-httpbis-priority.md

can reach at least equivalent performance characteristics compared to the more
complex HTTP/2 setups seen in practice, at least for the web use case.

## Disabling HTTP/2 Priorities {#settings-this-scheme}
The problems and insights set out above are motivation for allowing endpoints to
opt out of using the HTTP/2 priority scheme. Endpoints would benefit from
understanding their peer's intention, so the new
SETTINGS_DISABLE_HTTP2_PRIORITIES is defined. The value of the parameter MUST be
0 or 1.

If either side sends the parameter with a value of `1`, clients SHOULD NOT send
the HTTP/2 priority signal in HEADERS or PRIORITY frames. If both sides send the
parameter with a value of `0`, then both parties MAY use HTTP/2 priorities as
they see fit. A sender MUST NOT send the parameter with the value of `1` after
previously sending a value of `0`. If a client or server does not send the

> setting, the peer SHOULD assume that HTTP/2 priority signals are interpreted.

Reviewer: I think that the logic stated in this paragraph might be in reverse. It is my understanding that both peers use the H2 scheme on that hop, unless *both* endpoints agree on deprecation (i.e. `1`).

Therefore, something like:
* _A client SHOULD send the HTTP/2 priority signal until it receives the peer's SETTINGS_DISABLE_HTTP2_PRIORITIES settings with value of 1._
* _As the SETTINGS frame precedes any priority signal sent from a client, a server can determine if it should respect the H2 scheme or this prioritization scheme before building state_.

can reach at least equivalent performance characteristics compared to the more
complex HTTP/2 setups seen in practice, at least for the web use case.

## Disabling HTTP/2 Priorities
The problems and insights set out above are motivation for allowing endpoints to
opt out of using the HTTP/2 priority scheme. Endpoints would benefit from
understanding their peer's intention, so the new

> SETTINGS_DISABLE_HTTP2_PRIORITIES is defined. The value of the parameter MUST be

Reviewer: Would it be easier to describe if we change "DISABLE" in the name to "DEPRECATE"?

Author: yes

Reviewer: Done in 36b746b.

> 0 or 1.

Reviewer: How about adding something like: _This SETTINGS parameter MUST be sent as part of the first SETTINGS frame. When an endpoint receives the first SETTINGS frame, it learns if the peer has deprecated the HTTP/2 priority scheme, by consulting the value of SETTINGS_DISABLE_HTTP2_PRIORITY contained in that SETTINGS frame (or through the lack of that SETTINGS parameter)._

Reviewer: Done in 40e9875.

> An HTTP/2 client SHOULD use the HTTP/2 priority scheme (as it sees fit) until it

Reviewer: Before this paragraph, I think we might want to recommend the client to send both signals (i.e. the H2 scheme and the new scheme).

To paraphrase, I am suggesting of stating things in the order they happen:
1. client initially sends both
2. when the client learns H2 scheme is deprecated, the client stops sending H2 priorities
3. when the client learns H2 scheme is NOT deprecated, it _might_ stop sending Priority headers (though it can continue)

Reviewer: Done in 40e9875.

receives the peer's SETTINGS_DISABLE_HTTP2_PRIORITIES settings with the value of
`1`. As the SETTINGS frame precedes any priority signal sent from a client, a server can
determine if it should respect the HTTP/2 scheme before building state.

A client that disables the HTTP/2 priority scheme SHOULD subtitute it with some
other means of signaling request priority. The scheme extensible priority scheme

> is RECOMMENDED.

Reviewer: The client SHOULD use the extensible priority scheme until it finds out that the server has deprecated the HTTP/2 priority scheme. It MAY continue sending the Priority header field even after it learns that the server is using the HTTP/2 priority frame, as header fields are end-to-end signal.

Reviewer: Done in 40e9875.

can reach at least equivalent performance characteristics compared to the more
complex HTTP/2 setups seen in practice, at least for the web use case.

## Disabling HTTP/2 Priorities

The problems and insights set out above are motivation for allowing endpoints to
opt out of using the HTTP/2 priority scheme, in favor of using an alternative

> such as the scheme defined in this specification. Endpoints would benefit from

Reviewer 2: ```suggestion
such as the scheme defined in this specification. Endpoints benefit from
```

understanding their peer's intention, so the new
SETTINGS_DEPRECATE_HTTP2_PRIORITIES is defined. The value of the parameter MUST

> be 0 or 1.

Reviewer 2: What if it's not 0 or 1?

> Enpoints MUST send this SETTINGS parameter as part of the first SETTINGS frame.

Reviewer 2: ```suggestion
Enpoints MUST send this SETTINGS parameter in the first SETTINGS frame.
```

When the peer receives the first SETTINGS frame, it learns if the sender has
deprecated the HTTP/2 priority scheme, by consulting the value of

> SETTINGS_DEPRECATE_HTTP2_PRIORITIES parameter (or through lack of that

Reviewer 2: What does "(or through lack of that parameter)" mean, exactly?  I assume if it's not present, then it indicates HTTP/2 priorities are (at least possibly) supported?

Author: If a client and a server supports `SETTINGS_DEPRECATE_HTTP2_PRIORITIES` but a client omits the setting then the "initial value" of 0 is used (meaning that deprecate is not enabled). I also just added the requirement that once a setting is expressed it cannot be changed.

parameter).

Until the client receives the SETTINGS frame from the server, the client SHOULD
send both the priority signal defined in the HTTP/2 priority scheme (as it sees
fit) and also that of this prioritization scheme. Once the client learns that
the HTTP/2 priority scheme is deprecated, it SHOULD stop sending the HTTP/2
priority signals. If the client learns that the HTTP/2 priority scheme is not
deprecated, it SHOULD stop sending PRIORITY_UPDATE frames, but MAY continue
sending the Priority header field, as it is an end-to-end signal that might be
useful to nodes behind the server that the client is directly connected to.

> As the SETTINGS frame precedes any priority signal sent from a client, a server

Reviewer 2: This isn't completely true in HTTP/3 if SETTINGS are lost or delayed.

Reviewer: This is section is about deprecating _HTTP/2_ priorities, therefore we do not need to worry about H3.

Reviewer 2: Thanks

can reach at least equivalent performance characteristics compared to the more
complex HTTP/2 setups seen in practice, at least for the web use case.

## Disabling HTTP/2 Priorities

The problems and insights set out above are motivation for allowing endpoints to
opt out of using the HTTP/2 priority scheme, in favor of using an alternative
such as the scheme defined in this specification. Endpoints benefit from
understanding their peer's intention, so the new
SETTINGS_DEPRECATE_HTTP2_PRIORITIES is defined. The value of the parameter MUST
be 0 or 1.

Endpoints MUST send this SETTINGS parameter as part of the first SETTINGS frame.
When the peer receives the first SETTINGS frame, it learns if the sender has

> deprecated the HTTP/2 priority scheme, by consulting the value of

Reviewer 2: Suggestion "It learns the sender has deprecated the HTTP/2 priority scheme if it receives a value of other than 0."

can reach at least equivalent performance characteristics compared to the more
complex HTTP/2 setups seen in practice, at least for the web use case.

## Disabling HTTP/2 Priorities

The problems and insights set out above are motivation for allowing endpoints to

> opt out of using the HTTP/2 priority scheme, in favor of using an alternative

Reviewer 2: ```suggestion
disable the HTTP/2 priority scheme, in favor of using an alternative
```

such as the scheme defined in this specification. Endpoints benefit from
understanding their peer's intention, so the new

> SETTINGS_DEPRECATE_HTTP2_PRIORITIES is defined. The value of the parameter MUST

Reviewer 2: Suggestion: "SETTINGS_DEPRECATE_HTTP2_PRIORITIES enables a peer to understand their intention."

be 0 or 1. Any value other than 0 or 1 MUST be treated as a connection error
(see {{!RFC7540}}; Section 5.4.1) of type PROTOCOL_ERROR.

Endpoints MUST send this SETTINGS parameter as part of the first SETTINGS frame.
When the peer receives the first SETTINGS frame, it learns the sender has
deprecated the HTTP/2 priority scheme if it receives the
SETTINGS_DEPRECATE_HTTP2_PRIORITIES parameter with the value of 1.

A sender MUST NOT change the SETTINGS_DEPRECATE_HTTP2_PRIORITIES parameter value
after the first SETTINGS frame. Detection of a change by a receiver MUST be
treated as a connection error of type PROTOCOL_ERROR.

> Until the client receives the SETTINGS frame from the server, the client SHOULD

Reviewer 3: "as it sees fit" sounds like this should be a MAY

Reviewer 2: The parenthetical was removed in #116 but it's still a SHOULD, FWIW.

Reviewer: I think the current text (SHOULD without "as it sees fit") is fine.

We could change the text to something like: _"SHOULD ..., unless the client has the prior knowledge that the server lacks support for the HTTP/2 prioritization scheme, or that a different prioritization scheme will be used."_ But I am uncertain if we need to be that verbose.

the response as if it was assigned the priority of  `u=1, i=?1`
(i.e. round-robin) regardless of the value of the Priority header field being

> transmitted, unless the server has the knowledge that no intermediaries are

Reviewer 2: ```suggestion
transmitted, unless the server knows the intermediary is not
```
